File: algorithms/214.Shortest_Palindrome/Shortest_Palindrome.cpp
<<<<<<< HEAD
class Solution {

 public:
  string shortestPalindrome(string s) {
    if(s.size() == 0) return s;
    string extendS = s + "#" + string(s.rbegin(), s.rend());
    vector<int> prefix(extendS.size(), 0);
    for(int i = 1; i < extendS.size(); ++i) {
      int k = prefix[i-1];
      while(k > 0 && extendS[k] != extendS[i]) k = prefix[k-1];
      if(extendS[k] == extendS[i]) k += 1;
      prefix[i] = k;
    }
    string tmp = s.substr(prefix.back(), s.size() - prefix.back());
    return string(tmp.rbegin(), tmp.rend()) + s;
  }
=======
class Solution {

 public:
  string shortestPalindrome(string s) {
    if(s.size() == 0) return s;
    string extendS = s + "#" + string(s.rbegin(), s.rend());
    vector<int> prefix(extendS.size(), 0);
    for(int i = 1; i < extendS.size(); ++i) {
      int k = prefix[i-1];
      while(k > 0 && extendS[k] != extendS[i]) k = prefix[k-1];
      if(extendS[k] == extendS[i]) k += 1;
      prefix[i] = k;
    }
    string tmp = s.substr(prefix.back(), s.size() - prefix.back());
    return string(tmp.rbegin(), tmp.rend()) + s;
  }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
};

File: algorithms/214.Shortest_Palindrome/Shortest_Palindrome.java
public class Solution {

    public String shortestPalindrome(String s) {

        int j = 0;

        for (int i = s.length() - 1; i >= 0; i--) {//找到第一个使他不回文的位置

           if (s.charAt(i) == s.charAt(j)) { 

               j += 1; 

           }

        }

        if (j == s.length()) {  //本身是回文

            return s; 

        }

        String suffix = s.substring(j); // 后缀不能够匹配的字符串

        String prefix = new StringBuilder(suffix).reverse().toString(); // 前面补充prefix让他和suffix回文匹配

        String mid = shortestPalindrome(s.substring(0, j)); //递归调用找 [0,j]要最少可以补充多少个字符让他回文

        String ans = prefix + mid  + suffix;

        return  ans;

    }

}